nav2_waypoint_follower/plugins/photo_at_waypoint.cpp

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2020 Fetullah Atas
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_waypoint_follower/plugins/photo_at_waypoint.hpp"
+#include <pluginlib/class_list_macros.hpp>
+
+#include <string>
+#include <memory>
+
+
+namespace nav2_waypoint_follower
+{
+PhotoAtWaypoint::PhotoAtWaypoint()
+{
+}
+
+PhotoAtWaypoint::~PhotoAtWaypoint()
+{
+}
+
+void PhotoAtWaypoint::initialize(
+  const rclcpp_lifecycle::LifecycleNode::WeakPtr & parent,
+  const std::string & plugin_name)
+{
+  auto node = parent.lock();
+
+  node->declare_parameter(plugin_name + ".enabled", rclcpp::ParameterValue(true));
+  node->declare_parameter(
+    plugin_name + ".image_topic",
+    rclcpp::ParameterValue("/camera/color/image_raw"));
+  node->declare_parameter(plugin_name + ".save_dir", rclcpp::ParameterValue("/home/username/"));
+  node->declare_parameter(plugin_name + ".image_format", rclcpp::ParameterValue(".png"));
+
+  node->get_parameter(plugin_name + ".enabled", is_enabled_);
+  node->get_parameter(plugin_name + ".image_topic", image_topic_);
+  node->get_parameter(plugin_name + ".save_dir", directory_to_save_images_);
+  node->get_parameter(plugin_name + ".image_format", image_format_);
+
+  if (!is_enabled_) {
+    RCLCPP_INFO(
+      logger_, "Waypoint task executor plugin is disabled.");
+  } else {
+    RCLCPP_INFO(
+      logger_, "Initializing photo at waypoint plugin, subscribing to camera topic named; %s",
+      image_topic_.c_str());
+    camera_image_subscriber_ = node->create_subscription<sensor_msgs::msg::Image>(
+      image_topic_, rclcpp::SystemDefaultsQoS(),
+      std::bind(&PhotoAtWaypoint::imageCallback, this, std::placeholders::_1));
+  }
+}
+
+bool PhotoAtWaypoint::processAtWaypoint(
+  const geometry_msgs::msg::PoseStamped & curr_pose, const int & curr_waypoint_index)
+{
+  if (!is_enabled_) {
+    RCLCPP_WARN(
+      logger_,
+      "Photo at waypoint plugin is disabled. Not performing anything"
+    );
+    return true;
+  }
+  try {
+    std::experimental::filesystem::path save_dir = directory_to_save_images_;
+    std::experimental::filesystem::path file_name = std::to_string(
+      curr_waypoint_index) + "_" +
+      std::to_string(curr_pose.header.stamp.sec) + image_format_;
+
+    std::experimental::filesystem::path full_path_image_path = save_dir / file_name;
+
+
+    global_mutex_.lock();
```

Comment:
Use a lock_guard when you lock, it will automatically unlock once its out of scope. This is really fragile and if an exception is thrown, you don't unlock right now which will completely deadlock